nixos/modules/system/boot/networkd.nix

```diff
@@ -1045,44 +1053,49 @@ in
 
   };
 
-  config = mkIf config.systemd.network.enable {
+  config = mkMerge [
+    # .link units are honored by udev, no matter if systemd-networkd is enabled or not.
+    {
+      systemd.network.units = mapAttrs' (n: v: nameValuePair "${n}.link" (linkToUnit n v)) cfg.links;
```

Comment:
This has the implication of always renaming, reconfiguring, twiddling with, … interfaces even if the user doesn't care about networkd? 

This should/can be the default but we must at least provide a way to opt-out of this.

It might be the way udev wants to handle interface renaming but it might also lead to very unfortunate hard to debug behavior. It is certainly not expected if you are intimately familiar with both systemd and/or udev.

Reply:
Only if link units exist, which I think isn't the case unless they're created explicitly?

Reply:
`.link` files are parsed and acted upon by udev, not networkd. These are separate concepts, but currently the assembly of the `.link` files from `systemd.network.links` was only made if networkd wasn't enabled, which is the bug this PR is fixing.

People had to resort to dropping files in `/etc/systemd/network/*.link` manually, or write other custom udev rules to accomplish the same thing. (For some subsets of what's possible with `.link` files, there's a counterpart in scripted networking, manually shelling out to `ip link set mtu/(mac)address`)

Since some systemd versions, we (and other distros) ship `${pkgs.systemd}/lib/systemd/network/99-default.link` (which sets `MACAddressPolicy=persistent`) - this is also not conditional of networkd being enabled.

So back to the question - we still only rename, reconfigure, twiddle only if the user sets something in `systemd.networkd.links`, so I'd assume all configuration there is intentional - the `99-default.link` comes via another mechanism and is already enabled since systemd 242.

Reply:
My fear here is not that the *user* sets something but that we as distro (accidentially) ship new units but arguably that is a bug and should then be fixed upstream...

Reply:
So the only consumer here (for now) is the `50-zerotier` one - and this was already active, as it was added through `environment.etc."systemd/network/50-zerotier.link"` - just bypassing the (broken) `systemd.network.links` architecture, which this PR is meant to solve.